### PR TITLE
[BEAM-6491] Remove race in FileIOTest.testMatchWatchForNewFiles

### PR DESCRIPTION
**Pull request by jklukas**
This test was checking lastModificationTime for files that might not yet exist.
We now create the files beforehand so lastModificationTime is known, and have
the writer thread copy them into place, preserving the metadata.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by jklukas** [comment]
cc @iemejia @ajamato @jkff

**Comment by jklukas** [comment]
Run Java PreCommit

**Comment by jklukas** [comment]
First run of Java PreCommit failed `org.apache.beam.runners.dataflow.worker.util.MemoryMonitorTest.detectGCThrashing` which I assume is transient.

**Comment by iemejia** [comment]
[Postcommits look ok](https://builds.apache.org/job/beam_PreCommit_Java_Commit/3703/testReport/org.apache.beam.sdk.io/FileIOTest/testMatchWatchForNewFiles/).

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileIOTest.java [hunk 1]
```diff
@@ -221,36 +227,33 @@ public void testMatchWatchForNewFiles() throws IOException, InterruptedException
...
 
+    // We fetch lastModifiedTime from the files in the "source" directory to avoid a race condition
+    // with the writer thread.
     List<MatchResult.Metadata> expected =
         Arrays.asList(
             metadata(
-                basePath.resolve("first"),
-                42,
-                Files.getLastModifiedTime(basePath.resolve("first")).toMillis()),
+                watchPath.resolve("first"), 42, lastModifiedMillis(sourcePath.resolve("first"))),
```
Comment (ajamato) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileIOTest.java]: Is the lastModifiedTime changed when the sourcePath files are written? Why do we expect this match when the file are written back out a second time in the watchPath dir?

Or is the last modified time copied from the sourcePath files? So we expect it to be the same?

Would you please add a comment in this test clarifying this for the next reader?

Comment (jklukas) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileIOTest.java]: Pushed a commit to add additional commentary and to explicitly pass `StandardCopyOption.COPY_ATTRIBUTES` to signal that we are requesting attributes like timestamps to be preserved.

**Comment by ajamato** [comment]
I am not a committer though, would someone else please merge this for Jeff?

**Comment by iemejia** [comment]
Merged manually to squash, already pushed. Thanks @jklukas for the quick fix.

**Pull request closed without merging**